quantumblacklabs/kedro-viz
# [KED-2559] Add a new FastAPI-based server for viz

## Description

This PR makes the following changes:

#### 🏛 More modular, testable and extensible architecture

* Architecture [doc](https://github.com/quantumblacklabs/kedro-viz/blob/KED-2559/new-viz-backend/package/kedro_viz/docs/architecture.drawio) is checked into the codebase. It can be viewed with [draw.io](draw.io)
* Previously hard-to-test or untested code are now unit testable without having to setup an entire mock Kedro project.
* Swap out Flask for FastAPI ✨

#### 📚 API Documentation

* Automatic API documentation is available at [http://127.0.0.1:4141/docs/](http://127.0.0.1:4141/docs/) (Swagger UI) or [http://127.0.0.1:4141/redoc/](http://127.0.0.1:4141/redoc/) (Redoc UI). I like Redoc better.

![Screenshot 2021-05-14 at 01 01 02](https://user-images.githubusercontent.com/2032984/118201674-e73eca00-b44f-11eb-834c-ac77c1e105c5.png)

#### 🧪 Better testing

* More unit tests for previously hard-to-test or untested logic (there were quite a number of `pragma: no cover`), so we caught bugs like [this one](https://github.com/quantumblacklabs/kedro-viz/issues/433)
* Remove unnecessary / commented out e2e tests and simplify e2e test setup (a little bit). Will revisit.

#### 🧽 Hygiene

* Add `mypy` to `make pylint` 🥳
* Update legal header to 2021 and fix some formatting issues.
* Upgrade all dependencies (both `requirements.txt` and `test_requirements.txt`). That means we are on latest `pylint`, `flake8`, etc. and even a new `black`.
* Point linting tools to correct configuration files.


## How to review

Follow the components in the architecture to review each of them one by one.

![architecture](https://user-images.githubusercontent.com/2032984/118201082-98dcfb80-b44e-11eb-8e63-d7cc7254d1d1.png)

You can also review commit-by-commit. I have tried to make them self-contained(ish)


## QA notes

* You can try this out by checking out the branch and install the development version with `make install`
* You can also use the development server with

```
// in the root directory of the project
$ PYTHONPATH=$(pwd)/package python package/kedro_viz/server.py <path-to-your-kedro-project>
```

Note that you have to add `package` dir to `PYTHONPATH` here because `server.py` isn't self-contained anymore.


## Checklist


## Legal notice


- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
